### PR TITLE
core: Add support for List<EAG> in Subchannels

### PR DESCRIPTION
**Pull request by ejona86**
This avoids the needs to flatten to EAGs for cases like PickFirst,
making the Attributes in EAGs able to be used in communication with
core. See #4302 for some discussion on the topic.

-----

This alone will not be enough to rid ourselves of all flattening. We also
need to support List\<EAG> with OobChannels. But I'll do that as a later
PR.

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java [hunk 1]
```diff
@@ -703,4 +711,68 @@ public void closed(
...
+    }
+
+    /** Update to new groups, resetting the current index. */
+    public void updateGroups(List<EquivalentAddressGroup> newGroups) {
+      addressGroups = newGroups;
+      reset();
+    }
+
+    /** Returns false if the needle was not found and the current index was left unchanged. */
+    public boolean seekTo(SocketAddress needle) {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/InternalSubchannel.java]: Interesting argument naming :)

Comment (carl-mastrangelo) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: Possibly inspired by http://php.net/manual/en/function.strpos.php ?

File: core/src/main/java/io/grpc/LoadBalancer.java [hunk 2]
```diff
@@ -455,21 +456,61 @@ public boolean equals(Object other) {
...
+    /**
+     * Creates a Subchannel, which is a logical connection to the given group of addresses which are
+     * considered equivalent.  The {@code attrs} are custom attributes associated with this
+     * Subchannel, and can be accessed later through {@link Subchannel#getAttributes
+     * Subchannel.getAttributes()}.
+     *
+     * <p>The LoadBalancer is responsible for closing unused Subchannels, and closing all
+     * Subchannels within {@link #shutdown}.
+     *
+     * @throws IllegalArgumentException if {@code attrs} is empty
```
[the remark this answers is not in the document]
Comment (ejona86) [reply core/src/main/java/io/grpc/LoadBalancer.java]: Done. Very good catch. Amazingly good catch.

File: core/src/main/java/io/grpc/internal/InternalSubchannel.java (hunk 1 above)
Comment (ejona86) [reply core/src/main/java/io/grpc/internal/InternalSubchannel.java]: Quit trolling @carl-mastrangelo! :smile: Choose a more respectable source, like http://man7.org/linux/man-pages/man3/strstr.3.html !

**Pull request merged**